### PR TITLE
MAINT: Move ndarray.__str__ and ndarray.__repr__ to their own file

### PR DESCRIPTION
**Pull request by eric-wieser**
Goal here is to let the definition for `__unicode__` be in the same place too (#9201).

<s>This doesn't build, claiming there's a mismatch between the API dict and index. 

What have I done wrong?</s>

### CODE REVIEW

File: numpy/core/src/multiarray/strfuncs.c [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+
+static PyObject *PyArray_StrFunction = NULL;
+static PyObject *PyArray_ReprFunction = NULL;
+
+/*NUMPY_API
+ * Set the array print function to be a Python function.
+ */
+NPY_NO_EXPORT void
+PyArray_SetStringFunction(PyObject *op, int repr)
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/strfuncs.c]: Seems that this isn't being picked up by the tool to build the numpy API

Comment (mhvk) [inline numpy/core/src/multiarray/strfuncs.c]: You need to add things in `numpy/core/setup.py` -- `multiarray_deps`

Comment (mhvk) [inline numpy/core/src/multiarray/strfuncs.c]: Ah, just too late to help!

Comment (eric-wieser) [inline numpy/core/src/multiarray/strfuncs.c]: I have done already

Comment (eric-wieser) [inline numpy/core/src/multiarray/strfuncs.c]: Nope, help still needed - it still doesn't build for that reason

Comment (eric-wieser) [inline numpy/core/src/multiarray/strfuncs.c]: (That fixup was some other errors that I spotted that would appear once I fix this API one)

Comment (eric-wieser) [inline numpy/core/src/multiarray/strfuncs.c]: Soled - `API_FILES` in getapi.py

Comment (juliantaylor) [inline numpy/core/src/multiarray/strfuncs.c]: the `_deps` listings should be unnecessary now, nothing actually reads them and for gcc compatible compilers it determines dependencies automatically.

**Comment by eric-wieser** [comment]
All set, builds locally now

File: numpy/core/src/multiarray/strfuncs.c (hunk 1 above)
Comment (mhvk) [inline numpy/core/src/multiarray/strfuncs.c]: Wish these things were documented a bit more clearly... Raised issue gh-9203 to remind us.

**Comment by mhvk** [comment]
This looks all OK now. Hopefully the tests will bear this out...

**Comment by eric-wieser** [comment]
Tests are good!

**Comment by mhvk** [comment]
OK, even though this hasn't been out long enough for it to be necessarily noticed by all, I think this is a clear enough improvement that I'll just merge it. Then, #9201 can proceed as well.

**Pull request merged**